lib/node/NodeMainTemplatePlugin.js

```diff
@@ -2,130 +2,160 @@
-								return "\" + " + JSON.stringify(shortChunkHashMap) + "[chunkId] + \"";
-							},
-							name: "\" + (" + JSON.stringify(chunkMaps.name) + "[chunkId]||chunkId) + \""
-						}
-					}) + ";",
-					"require('fs').readFile(filename, 'utf-8',  function(err, content) {",
+const Template = require("../Template");
+
+module.exports = class NodeMainTemplatePlugin {
+	constructor(asyncChunkLoading) {
+		this.asyncChunkLoading = asyncChunkLoading;
+	}
+
+	apply(mainTemplate) {
+		const asyncChunkLoading = this.asyncChunkLoading;
+		mainTemplate.plugin("local-vars", function(source, chunk) {
+			if(chunk.chunks.length > 0) {
+				return this.asString([
+					source,
+					"",
+					"// object to store loaded chunks",
+					"// \"0\" means \"already loaded\"",
+					"var installedChunks = {",
+					this.indent(chunk.ids.map((id) => id + ": 0").join(",\n")),
+					"};"
+				]);
+			}
+			return source;
+		});
+		mainTemplate.plugin("require-extensions", function(source, chunk) {
+			if(chunk.chunks.length > 0) {
+				return this.asString([
+					source,
+					"",
+					"// uncatched error handler for webpack runtime",
+					this.requireFn + ".oe = function(err) {",
 					this.indent([
-						"if(err) return reject(err);",
-						"var chunk = {};",
-						"require('vm').runInThisContext('(function(exports, require, __dirname, __filename) {' + content + '\\n})', filename)" +
-						"(chunk, require, require('path').dirname(filename), filename);"
-					].concat(insertMoreModules).concat([
-						"var callbacks = [];",
-						"for(var i = 0; i < chunkIds.length; i++) {",
+						"process.nextTick(function() {",
+						this.indent("throw err; // catch this error by using System.import().catch()"),
+						"});"
+					]),
+					"};"
+				]);
+			}
+			return source;
+		});
+		mainTemplate.plugin("require-ensure", function(_, chunk, hash) {
+			const chunkFilename = this.outputOptions.chunkFilename;
+			const chunkMaps = chunk.getChunkMaps();
+			const insertMoreModules = [
+				"var moreModules = chunk.modules, chunkIds = chunk.ids;",
+				"for(var moduleId in moreModules) {",
+				this.indent(this.renderAddModule(hash, chunk, "moduleId", "moreModules[moduleId]")),
+				"}"
+			];
+			if(asyncChunkLoading) {
+				return this.asString([
+					"// \"0\" is the signal for \"already loaded\"",
+					"if(installedChunks[chunkId] === 0)",
+					this.indent([
+						"return Promise.resolve();"
+					]),
+					"// array of [resolve, reject, promise] means \"currently loading\"",
+					"if(installedChunks[chunkId])",
+					this.indent([
+						"return installedChunks[chunkId][2];"
+					]),
+					"// load the chunk and return promise to it",
+					"var promise = new Promise(function(resolve, reject) {",
+					this.indent([
+						"installedChunks[chunkId] = [resolve, reject];",
+						"var filename = __dirname + " + this.applyPluginsWaterfall("asset-path", JSON.stringify("/" + chunkFilename), {
+							hash: "\" + " + this.renderCurrentHashCode(hash) + " + \"",
```

Comment:
This part of [code](https://github.com/webpack/webpack/blob/master/lib/node/NodeMainTemplatePlugin.js#L73-L91) is repeated two more times in this file - [here](https://github.com/webpack/webpack/blob/master/lib/node/NodeMainTemplatePlugin.js#L118-L136) and [here](https://github.com/webpack/webpack/blob/master/lib/node/NodeMainTemplatePlugin.js#L155-L173).
Maybe I should try to refactor it out? Or is it not in the scope of ES6 refactoring?